pkg/kubernetes/watcher.go

```diff
@@ -225,39 +230,46 @@ func (w *ObjListWatcher) Stop() {
 	close(w.stopChannel)
 }
 
-func (w *ObjListWatcher) handleAdd(obj interface{}) {
+func (w *ObjListWatcher) handleAdd(obj interface{}) error {
 	switch w.ResourceKind {
 	case podResourceType:
 		pod, ok := obj.(*corev1.Pod)
 		if !ok {
-			klog.Infof("Could not convert obj: %v", w.ResourceKind)
-			return
+			err := fmt.Errorf("could not convert obj: %v", w.ResourceKind)
+			return err
 		}
 		for _, condition := range pod.Status.Conditions {
 			if condition.Type != corev1.ContainersReady {
 				continue
 			}
 			klog.V(5).Infof("Pod %s %s is ready with %d container statuses, %d init container status, %d ephemeral statues",
 				pod.Name, pod.Namespace, len(pod.Status.ContainerStatuses), len(pod.Status.InitContainerStatuses), len(pod.Status.EphemeralContainerStatuses))
-			w.Mx.Lock()
-			err1 := w.fillInfo(pod, pod.Status.ContainerStatuses)
-			err2 := w.fillInfo(pod, pod.Status.InitContainerStatuses)
-			err3 := w.fillInfo(pod, pod.Status.EphemeralContainerStatuses)
-			w.Mx.Unlock()
-			klog.V(5).Infof("parsing pod %s %s status: %v %v %v", pod.Name, pod.Namespace, err1, err2, err3)
+			if err := w.fillInfo(pod, pod.Status.ContainerStatuses, w.Mx); err != nil {
```

Comment:
```go
for statuses := range [pod.Status.ContainersStatuses, pod.Status.InitContainerStatuses, pod.Status.EphemeralContainerStatuses] {
    if err := if err := w.fillInfo(pod, statuses, w.Mx); err !=nil {
        return err
    }
}
```

Actually we could go one better and just concatenate those arrays to avoid iterating again inside `fillInfo`

Reply:
I will tidy it up on Monday morning